plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs

```diff
@@ -0,0 +1,243 @@
+GHC. This stops GHC from inlining it. As a bonus, it also won't be inlined if
+that function is compiled later into the body of another function.
+
+We do therefore need to handle 'noinline' in the plugin, as it itself does not have
+an unfolding.
+
+Another annoying quirk: even if you have 'noinline'd a function call, if the body is
+a single variable, it will still inline! This is the case for the obvious definition
+of 'stringToBuiltinString' (since the newtype constructor vanishes), so we have to add
+some obfuscation to the body to prevent it inlining.
+-}
+
+obfuscatedId :: a -> a
+obfuscatedId a = a
+{-# NOINLINE obfuscatedId #-}
+
+stringToBuiltinByteString :: Haskell.String -> BuiltinByteString
+stringToBuiltinByteString str = encodeUtf8 $ stringToBuiltinString str
+{-# INLINABLE stringToBuiltinByteString #-}
+
+stringToBuiltinString :: Haskell.String -> BuiltinString
+-- To explain why the obfuscatedId is here
+-- See Note [noinline hack]
+stringToBuiltinString str = obfuscatedId (BuiltinString $ Text.pack str)
+{-# INLINABLE stringToBuiltinString #-}
+
+{- Same noinline hack as with `String` type. -}
+instance IsString BuiltinByteString where
+    -- Try and make sure the dictionary selector goes away, it's simpler to match on
+    -- the application of 'stringToBuiltinByteString'
+    -- See Note [noinline hack]
+    fromString = Magic.noinline stringToBuiltinByteString
+    {-# INLINE fromString #-}
+
+-- We can't put this in `Builtins.hs`, since that force `O0` deliberately, which prevents
+-- the unfoldings from going in. So we just stick it here. Fiddly.
+instance IsString BuiltinString where
+    -- Try and make sure the dictionary selector goes away, it's simpler to match on
+    -- the application of 'stringToBuiltinString'
+    -- See Note [noinline hack]
+    fromString = Magic.noinline stringToBuiltinString
+    {-# INLINE fromString #-}
+
+{- Note [Built-in types and their Haskell versions]
+'HasToBuiltin' allows us to convert a value of a built-in type such as 'ByteString' to its Plutus
+Tx counterpart, 'BuiltinByteString' in this case. The idea is the same for all built-in types: just
+take the Haskell version and make it the Plutus Tx one.
+
+'HasToOpaque' is different, we use it for converting values of only those built-in types that exist
+in the Plutus Tx realm. I.e. we cannot convert a 'ByteString', since 'ByteString's don't exist in
+Plutus Tx, only 'BuiltinByteString's do.
+
+Consider, say, the built-in pair type. In Plutus Tx, we have an (opaque) type for this. It's opaque
+because you can't actually pattern match on it, instead you can only in fact use the specific
+functions that are available as builtins.
+
+We _also_ have the normal Haskell pair type. This is very different: you can
+pattern match on it, and you can use whatever user-defined functions you like on it.
+
+Users would really like to use the latter, and not the former. So we often want
```

Comment:
> Users would really like to use the latter, and not the former.

It would be nice to expand on why users would like it.

Reply:
That part was authored by @michaelpj. I assume because pattern matching is more ergonomic than calling builtins manually, although we could have pattern matching synonyms, although those would be slower and in a very obscure-to-the-user way.

Reply:
Ah, well, you also have all the library functions for transparent types, like `map`, `foldr`, `concat` etc for lists, none of the is available for `BuiltinList`.